components/shared/ui/appPage.tsx

```diff
@@ -5,31 +5,79 @@ import { perCollectionSEOTitle } from "../../../lib/constants/labels";
 import { ValidCollectionCodename } from "../../../lib/types/perCollection";
 import { useSmartLink } from "../../../lib/useSmartLink";
 import { createItemSmartLink } from "../../../lib/utils/smartLinkUtils";
-import { Block_Navigation } from "../../../models";
+import { Article, Block_Navigation, Product, SEOMetadata, WSL_WebSpotlightRoot } from "../../../models";
 import { SiteCodenameProvider } from "../siteCodenameContext";
 import { Footer } from "./footer";
 import { Menu } from "./menu";
 
 type Props = Readonly<{
   children: ReactNode;
   siteCodename: ValidCollectionCodename;
-  itemId?: string;
+  item?: WSL_WebSpotlightRoot | Article | Product;
   siteMenu?: Block_Navigation;
+  defaultMetadata?: SEOMetadata;
+  pageType: "Article" | "WebPage" | "Product";
 }>;
 
 export const AppPage: FC<Props> = props => {
   useSmartLink();
+
+  const getPageTitle = (): string => {
+    const siteTitle = perCollectionSEOTitle[props.siteCodename];
+    const titleExtension = props.item?.elements.seoMetadataTitle.value;
+
+    return titleExtension && titleExtension !== siteTitle ? `${siteTitle} | ${titleExtension}` : siteTitle;
+  };
+
+  const pageMetaDescription = props.item?.elements.seoMetadataDescription.value ?? props.defaultMetadata?.elements.seoMetadataDescription.value;
+  const pageMetaKeywords = props.item?.elements.seoMetadataKeywords.value ?? props.defaultMetadata?.elements.seoMetadataKeywords.value;
+  const pageMetaTitle = getPageTitle();
+
   return (
     <SiteCodenameProvider siteCodename={props.siteCodename}>
       <Head>
-        <title>{perCollectionSEOTitle[props.siteCodename]}</title>
+        <title>{pageMetaTitle}</title>
+        <meta
```

Comment:
I'd suggest extracting the SEO-related rendering (and logic) into a separate component to make this one a bit more concise.

Reply:
extracted into separate component, passed props from parent